File: problem11/julia/matrix_product.jl
module MatrixProduct

export greatest_product

function greatest_product(matrix::Array{Int,2}, adj::Int)
    rows, cols = size(matrix)
    max_prod = 0
    best_coords = Vector{Tuple{Int,Int}}()
    for i in 1:rows
        for j in 1:cols
            # right
            if j + adj - 1 <= cols
                prod = 1
                coords = Vector{Tuple{Int,Int}}()
                for k in 0:adj-1
                    prod *= matrix[i, j+k]
                    push!(coords, (i, j+k))
                end
                if prod > max_prod
                    max_prod = prod
                    best_coords = coords
                end
            end
            # down
            if i + adj - 1 <= rows
                prod = 1
                coords = Vector{Tuple{Int,Int}}()
                for k in 0:adj-1
                    prod *= matrix[i+k, j]
                    push!(coords, (i+k, j))
                end
                if prod > max_prod
                    max_prod = prod
                    best_coords = coords
                end
            end
            # diag down-right
            if i + adj - 1 <= rows && j + adj - 1 <= cols
                prod = 1
                coords = Vector{Tuple{Int,Int}}()
                for k in 0:adj-1
                    prod *= matrix[i+k, j+k]
                    push!(coords, (i+k, j+k))
                end
                if prod > max_prod
                    max_prod = prod
                    best_coords = coords
                end
            end
            # diag down-left
            if i + adj - 1 <= rows && j - adj + 1 >= 1
                prod = 1
                coords = Vector{Tuple{Int,Int}}()
                for k in 0:adj-1
                    prod *= matrix[i+k, j-k]
                    push!(coords, (i+k, j-k))
                end
                if prod > max_prod
                    max_prod = prod
                    best_coords = coords
                end
            end
        end
    end
    return max_prod, best_coords
end

end # module

File: problem11/kotlin/src/main/kotlin/MatrixProduct.kt
package org.euler.problem11

fun greatestProduct(matrix: List<List<Int>>, adj: Int): Pair<Int, List<Pair<Int, Int>>> {
    val rows = matrix.size
    val cols = if (rows > 0) matrix[0].size else 0
    var max = 0
    var bestCoords = listOf<Pair<Int, Int>>()
    for (i in 0 until rows) {
        for (j in 0 until cols) {
            // right
            if (j + adj <= cols) {
                var prod = 1
                val coords = mutableListOf<Pair<Int, Int>>()
                for (k in 0 until adj) {
                    prod *= matrix[i][j + k]
                    coords.add(Pair(i, j + k))
                }
                if (prod > max) {
                    max = prod
                    bestCoords = coords.toList()
                }
            }
            // down
            if (i + adj <= rows) {
                var prod = 1
                val coords = mutableListOf<Pair<Int, Int>>()
                for (k in 0 until adj) {
                    prod *= matrix[i + k][j]
                    coords.add(Pair(i + k, j))
                }
                if (prod > max) {
                    max = prod
                    bestCoords = coords.toList()
                }
            }
            // diag down-right
            if (i + adj <= rows && j + adj <= cols) {
                var prod = 1
                val coords = mutableListOf<Pair<Int, Int>>()
                for (k in 0 until adj) {
                    prod *= matrix[i + k][j + k]
                    coords.add(Pair(i + k, j + k))
                }
                if (prod > max) {
                    max = prod
                    bestCoords = coords.toList()
                }
            }
            // diag down-left
            if (i + adj <= rows && j - adj + 1 >= 0) {
                var prod = 1
                val coords = mutableListOf<Pair<Int, Int>>()
                for (k in 0 until adj) {
                    prod *= matrix[i + k][j - k]
                    coords.add(Pair(i + k, j - k))
                }
                if (prod > max) {
                    max = prod
                    bestCoords = coords.toList()
                }
            }
        }
    }
    return Pair(max, bestCoords)
}
